core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -255,13 +254,8 @@ private object RTS {
           case a: IO.Attempt[_, _, _, _] =>
             errorHandler = a.err.asInstanceOf[Any => IO[Any, Any]]
           case f0: Finalizer[_] =>
-            val f = f0.asInstanceOf[Finalizer[E]]
-
-            // Lazy initialization of body:
-            if (body eq null) body = ExitResult.Failed(err)
-
-            val currentFinalizer: IO[E2, List[Throwable]] = f.finalizer(body).run.map(collectDefect)
-
+            val f                                           = f0.asInstanceOf[Finalizer[E]]
```

Comment:
This needs to be changed:

`f0.finalizer`

You only need to run the finalizer, not the whole wrapper action (`finalizer *> IO.now(a)`), which will decrease performance.